Increase specificity of mothur.pair.dist sniffer

bernt-matthias opened the pull request:
The datatype is a 3 column table where the last column must be numeric and contain at least one float. I think this is to unspecific - in particular because the sniffer seems to be one of the first.

First commit adds a test showing the problem.

Options:

- [ ] its OK .. leave as it is
- [ ] remove sniffer / change sniffing order (I guess changing the sniffing order would render the sniffer function useless anyway)

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
show that PairwiseDistanceMatrix sniffer

mis-sniffs tabular data because its to unspecific

bernt-matthias committed:
disambiguate test data (2.tabular)

nsoranzo commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -491,6 +491,9 @@ def sniff_prefix(self, file_prefix):
         >>> fname = get_test_fname( 'mothur_datatypetest_false.mothur.pair.dist' )
         >>> PairwiseDistanceMatrix().sniff( fname )
```
```suggestion
        >>> fname = get_test_fname('mothur_datatypetest_false.mothur.pair.dist')
        >>> PairwiseDistanceMatrix().sniff(fname)
```

nsoranzo commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -491,6 +491,9 @@ def sniff_prefix(self, file_prefix):
         >>> PairwiseDistanceMatrix().sniff( fname )
         False
+        >>> fname = get_test_fname( '2.tabular' )
+        >>> PairwiseDistanceMatrix().sniff( fname )
```
```suggestion
        >>> fname = get_test_fname('2.tabular')
        >>> PairwiseDistanceMatrix().sniff(fname)
```

nsoranzo reviewed (commented):
Maybe we could add a condition that the first and second columns should not all be numbers?

bernt-matthias committed:
Apply suggestions from code review

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

bernt-matthias committed:
invert check to reduce indentation

bernt-matthias committed:
stricted sniffer for PairwiseDistanceMatrix

bernt-matthias committed:
stricter test for PairwiseDistanceMatrix

check if columns 1 and 2 contain at least one non-float (text) entry

bernt-matthias committed:
stricter test for DistanceMatrix sniffer

could be implemented for all, but maybe if done in doctests
then to many sniff(filex) calls run multiple times.

bernt-matthias committed:
revert some changes

now included in https://github.com/galaxyproject/galaxy/pull/12418

nsoranzo commented on lib/galaxy/datatypes/sniff.py:
```diff
@@ -38,6 +38,24 @@ def get_test_fname(fname):
+    pos = set()
+    neg = set()
+    path, name = os.path.split(__file__)
+    path = os.path.join(path, 'test')
```
```suggestion
    path = os.path.join(os.path.dirname(__file__), 'test')
```

and I would probably move it outside of this function as a constant instead of recalculating it over and over. It can be reused by `get_test_fname()` as well.

bernt-matthias replied:
I implemented this is https://github.com/galaxyproject/galaxy/pull/12418 .. will remove this change here

nsoranzo commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -484,31 +484,42 @@ def sniff_prefix(self, file_prefix):
-        >>> PairwiseDistanceMatrix().sniff( fname )
+        >>> from galaxy.datatypes.sniff import get_test_iter
+        >>> pos, neg = get_test_iter(['mothur_datatypetest_true.mothur.pair.dist'])
+        >>> all([PairwiseDistanceMatrix().sniff(p) for p in pos])
```
```suggestion
        >>> all(PairwiseDistanceMatrix().sniff(p) for p in pos)
```

nsoranzo commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -484,31 +484,42 @@ def sniff_prefix(self, file_prefix):
+        >>> all([PairwiseDistanceMatrix().sniff(p) for p in pos])
+        True
+        >>> all([not PairwiseDistanceMatrix().sniff(n) for n in neg])
         True
```
```suggestion
        >>> any(PairwiseDistanceMatrix().sniff(n) for n in neg)
        False
```

bernt-matthias committed:
restore simple doc test
